File: src/main/java/org/ldk/structs/ProbeSendFailure.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;


/**
 * Indicates that we failed to send a payment probe. Further errors may be surfaced later via
 * [`Event::ProbeFailed`].
 * 
 * [`Event::ProbeFailed`]: crate::events::Event::ProbeFailed
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class ProbeSendFailure extends CommonBase {
	private ProbeSendFailure(Object _dummy, long ptr) { super(ptr); }
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		super.finalize();
		if (ptr != 0) { bindings.ProbeSendFailure_free(ptr); }
	}
	static ProbeSendFailure constr_from_ptr(long ptr) {
		bindings.LDKProbeSendFailure raw_val = bindings.LDKProbeSendFailure_ref_from_ptr(ptr);
		if (raw_val.getClass() == bindings.LDKProbeSendFailure.RouteNotFound.class) {
			return new RouteNotFound(ptr, (bindings.LDKProbeSendFailure.RouteNotFound)raw_val);
		}
		if (raw_val.getClass() == bindings.LDKProbeSendFailure.ParameterError.class) {
			return new ParameterError(ptr, (bindings.LDKProbeSendFailure.ParameterError)raw_val);
		}
		if (raw_val.getClass() == bindings.LDKProbeSendFailure.DuplicateProbe.class) {
			return new DuplicateProbe(ptr, (bindings.LDKProbeSendFailure.DuplicateProbe)raw_val);
		}
		assert false; return null; // Unreachable without extending the (internal) bindings interface
	}

	/**
	 * We were unable to find a route to the destination.
	 */
	public final static class RouteNotFound extends ProbeSendFailure {
		private RouteNotFound(long ptr, bindings.LDKProbeSendFailure.RouteNotFound obj) {
			super(null, ptr);
		}
	}
	/**
	 * A parameter which was passed to [`ChannelManager::send_probe`] was invalid, preventing us from
	 * attempting to send the probe at all.
	 * 
	 * You can freely resend the probe (with the parameter error fixed).
	 * 
	 * Because the probe failed outright, no payment tracking is done and no
	 * [`Event::ProbeFailed`] events will be generated.
	 * 
	 * [`ChannelManager::send_probe`]: crate::ln::channelmanager::ChannelManager::send_probe
	 * [`Event::ProbeFailed`]: crate::events::Event::ProbeFailed
	 */
	public final static class ParameterError extends ProbeSendFailure {
		public final org.ldk.structs.APIError parameter_error;
		private ParameterError(long ptr, bindings.LDKProbeSendFailure.ParameterError obj) {
			super(null, ptr);
			long parameter_error = obj.parameter_error;
			org.ldk.structs.APIError parameter_error_hu_conv = org.ldk.structs.APIError.constr_from_ptr(parameter_error);
			if (parameter_error_hu_conv != null) { parameter_error_hu_conv.ptrs_to.add(this); };
			this.parameter_error = parameter_error_hu_conv;
		}
	}
	/**
	 * Indicates that a payment for the provided [`PaymentId`] is already in-flight and has not
	 * yet completed (i.e. generated an [`Event::ProbeSuccessful`] or [`Event::ProbeFailed`]).
	 * 
	 * [`PaymentId`]: crate::ln::channelmanager::PaymentId
	 * [`Event::ProbeSuccessful`]: crate::events::Event::ProbeSuccessful
	 * [`Event::ProbeFailed`]: crate::events::Event::ProbeFailed
	 */
	public final static class DuplicateProbe extends ProbeSendFailure {
		private DuplicateProbe(long ptr, bindings.LDKProbeSendFailure.DuplicateProbe obj) {
			super(null, ptr);
		}
	}
	long clone_ptr() {
		long ret = bindings.ProbeSendFailure_clone_ptr(this.ptr);
		Reference.reachabilityFence(this);
		return ret;
	}

	/**
	 * Creates a copy of the ProbeSendFailure
	 */
	public ProbeSendFailure clone() {
		long ret = bindings.ProbeSendFailure_clone(this.ptr);
		Reference.reachabilityFence(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(this); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new RouteNotFound-variant ProbeSendFailure
	 */
	public static ProbeSendFailure route_not_found() {
		long ret = bindings.ProbeSendFailure_route_not_found();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new ParameterError-variant ProbeSendFailure
	 */
	public static ProbeSendFailure parameter_error(org.ldk.structs.APIError a) {
		long ret = bindings.ProbeSendFailure_parameter_error(a.ptr);
		Reference.reachabilityFence(a);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new DuplicateProbe-variant ProbeSendFailure
	 */
	public static ProbeSendFailure duplicate_probe() {
		long ret = bindings.ProbeSendFailure_duplicate_probe();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Checks if two ProbeSendFailures contain equal inner contents.
	 * This ignores pointers and is_owned flags and looks at the values in fields.
	 */
	public boolean eq(org.ldk.structs.ProbeSendFailure b) {
		boolean ret = bindings.ProbeSendFailure_eq(this.ptr, b.ptr);
		Reference.reachabilityFence(this);
		Reference.reachabilityFence(b);
		return ret;
	}

	@Override public boolean equals(Object o) {
		if (!(o instanceof ProbeSendFailure)) return false;
		return this.eq((ProbeSendFailure)o);
	}
}

File: c_sharp/src/org/ldk/structs/ProbeSendFailure.cs
using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {

/**
 * Indicates that we failed to send a payment probe. Further errors may be surfaced later via
 * [`Event::ProbeFailed`].
 * 
 * [`Event::ProbeFailed`]: crate::events::Event::ProbeFailed
 */
public class ProbeSendFailure : CommonBase {
	protected ProbeSendFailure(object _dummy, long ptr) : base(ptr) { }
	~ProbeSendFailure() {
		if (ptr != 0) { bindings.ProbeSendFailure_free(ptr); }
	}

	internal static ProbeSendFailure constr_from_ptr(long ptr) {
		long raw_ty = bindings.LDKProbeSendFailure_ty_from_ptr(ptr);
		switch (raw_ty) {
			case 0: return new ProbeSendFailure_RouteNotFound(ptr);
			case 1: return new ProbeSendFailure_ParameterError(ptr);
			case 2: return new ProbeSendFailure_DuplicateProbe(ptr);
			default:
				throw new ArgumentException("Impossible enum variant");
		}
	}

	/** A ProbeSendFailure of type RouteNotFound */
	public class ProbeSendFailure_RouteNotFound : ProbeSendFailure {
		internal ProbeSendFailure_RouteNotFound(long ptr) : base(null, ptr) {
		}
	}
	/** A ProbeSendFailure of type ParameterError */
	public class ProbeSendFailure_ParameterError : ProbeSendFailure {
		public org.ldk.structs.APIError parameter_error;
		internal ProbeSendFailure_ParameterError(long ptr) : base(null, ptr) {
			long parameter_error = bindings.LDKProbeSendFailure_ParameterError_get_parameter_error(ptr);
			org.ldk.structs.APIError parameter_error_hu_conv = org.ldk.structs.APIError.constr_from_ptr(parameter_error);
			if (parameter_error_hu_conv != null) { parameter_error_hu_conv.ptrs_to.AddLast(this); };
			this.parameter_error = parameter_error_hu_conv;
		}
	}
	/** A ProbeSendFailure of type DuplicateProbe */
	public class ProbeSendFailure_DuplicateProbe : ProbeSendFailure {
		internal ProbeSendFailure_DuplicateProbe(long ptr) : base(null, ptr) {
		}
	}
	internal long clone_ptr() {
		long ret = bindings.ProbeSendFailure_clone_ptr(this.ptr);
		GC.KeepAlive(this);
		return ret;
	}

	/**
	 * Creates a copy of the ProbeSendFailure
	 */
	public org.ldk.structs.ProbeSendFailure clone() {
		long ret = bindings.ProbeSendFailure_clone(this.ptr);
		GC.KeepAlive(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(this); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new RouteNotFound-variant ProbeSendFailure
	 */
	public static org.ldk.structs.ProbeSendFailure route_not_found() {
		long ret = bindings.ProbeSendFailure_route_not_found();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new ParameterError-variant ProbeSendFailure
	 */
	public static org.ldk.structs.ProbeSendFailure parameter_error(org.ldk.structs.APIError a) {
		long ret = bindings.ProbeSendFailure_parameter_error(a.ptr);
		GC.KeepAlive(a);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new DuplicateProbe-variant ProbeSendFailure
	 */
	public static org.ldk.structs.ProbeSendFailure duplicate_probe() {
		long ret = bindings.ProbeSendFailure_duplicate_probe();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.ProbeSendFailure ret_hu_conv = org.ldk.structs.ProbeSendFailure.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Checks if two ProbeSendFailures contain equal inner contents.
	 * This ignores pointers and is_owned flags and looks at the values in fields.
	 */
	public bool eq(org.ldk.structs.ProbeSendFailure b) {
		bool ret = bindings.ProbeSendFailure_eq(this.ptr, b.ptr);
		GC.KeepAlive(this);
		GC.KeepAlive(b);
		return ret;
	}

	public override bool Equals(object o) {
		if (!(o is ProbeSendFailure)) return false;
		return this.eq((ProbeSendFailure)o);
	}
}
} } }
